accord-core/src/main/java/accord/local/CommandStores.java

```diff
@@ -87,44 +88,30 @@ private static class StoreSupplier
             this.shardFactory = shardFactory;
         }
 
-        CommandStore create(int id, RangesForEpochHolder rangesForEpoch)
+        CommandStore create(int id, EpochUpdateHolder rangesForEpoch)
         {
             return shardFactory.create(id, time, agent, this.store, progressLogFactory, rangesForEpoch);
         }
     }
 
-    public static class RangesForEpochHolder
-    {
-        // no need for safe publication; RangesForEpoch members are final, and will be guarded by other synchronization actions
-        protected RangesForEpoch current;
-
-        /**
-         * This is updated asynchronously, so should only be fetched between executing tasks;
-         * otherwise the contents may differ between invocations for the same task.
-         * @return the current RangesForEpoch
-         */
-        public RangesForEpoch get() { return current; }
-    }
-
     static class ShardHolder
     {
         final CommandStore store;
-        final RangesForEpochHolder ranges;
+        RangesForEpoch ranges;
```

Comment:
The concern is that the values may not be current not that they will be invalid (since `RangesForEpoch` is all final fields)?

While the values can be a little stale there is a synchronizing action I believe in the form of the volatile read of the `current` field which does updated every time we also update a `RangesForEpoch`. 

I believe that might be a sufficient, but we are getting into the fancier stuff.

So what might happen is not so much that you read a stale value, but for a given `Shard` you could end up reading a newer value than matches the rest of the `Snapshot`.

Reply:
> The concern is that the values may not be current not that they will be invalid 

I don't think this is a regression as it moves the mutation from `RangesForEpochHolder` to `ShardHolder` and keeps the same access pattern... since everything is `final` the read concern is only for stale data.

Reply:
The idea actually was that we would avoid changing the data we were looking at for an in-flight operation. But now that I look at this I think we need a stronger guarantee: we need to ensure command processing is deterministic for the work @iamaleksey is doing, so we need to ensure that anything we ask for here is the same as would be returned with *more* information. But that's slightly off-topic. It is very deliberate that we return "stale" information for a command that began processing with that information to hand.